src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs

```diff
@@ -228,6 +228,48 @@ public IEnumerable<ZipFilePair> EnumeratePackageEntries(IEnumerable<string> pack
             }
         }
 
+        /// <summary>
+        /// Validate all files in package are not traversed outside of the expected extraction path.
+        /// Eg: file entry like ../../foo.dll can get outside of the expected extraction path.
+        /// </summary>
+        public async Task ValidatePackageEntriesAsync(CancellationToken token)
+        {
+            try
+            {
+                var files = await GetFilesAsync(token);
+                // This dummy destination is used to check if the file in package get outside of the extractionPath
+                var dummyDestination = NuGetEnvironment.GetFolderPath(NuGetFolderPath.NuGetHome);
+
+                if (!dummyDestination.EndsWith(Path.DirectorySeparatorChar.ToString()))
+                    dummyDestination += Path.DirectorySeparatorChar;
+
+
+                var result = files.All(p =>
+                {
+                    var normalizedPath = Uri.UnescapeDataString(p.Replace('/', Path.DirectorySeparatorChar));
+                    var fullPath = Path.GetFullPath(Path.Combine(dummyDestination, normalizedPath));
+                    return fullPath.StartsWith(dummyDestination, StringComparison.Ordinal);
```

Comment:
You should also consider whether you are ok with the destination path is the root directory. If not, then you also want to reject paths like `"."` by checking that the lengths are not the same.

Reply:
the destination path is never the root directory, we appended ".nuget" at the end of userProfile path.

Reply:
I think it's OK to have "." in the ZipArchive, right? Is there any concern about that case?

Reply:
Only in that you try to write a file to the same path as the destination directory. That will throw, of course (_assuming_ you have already created the folder). Unless you have a reason to not block it, I would.

Reply:
I see, block it